radio/wifi/wifi.go

```diff
@@ -0,0 +1,185 @@
+package wifi
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os/exec"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/lair-framework/go-nmap"
+	"github.com/parnurzeal/gorequest"
+)
+
+type Host struct {
+	name string
+	ip   string
+	mac  string
+}
+
+func StartHotspot() error {
+	if err := activateConnection("resin-hotspot"); err != nil {
+		return err
+	}
+
+	// Give the wifi devices a chance to connect
+	time.Sleep(5 * time.Second)
+	return nil
+}
+
+func StopHotspot() error {
+	return activateConnection("resin-wifi")
+}
+
+func Scan(id string) (map[string]struct{}, error) {
+	hosts, err := scan()
+	if err != nil {
+		return nil, err
+	}
+
+	online := make(map[string]struct{})
+	for _, host := range hosts {
+		if host.name == id {
+			var s struct{}
+			online[host.mac] = s
+		}
+	}
+
+	return online, nil
+}
+
+func Online(id string) (bool, error) {
+	hosts, err := scan()
+	if err != nil {
+		return false, err
+	}
+
+	for _, host := range hosts {
+		if host.mac == id {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+func GetIP(id string) (string, error) {
+	hosts, err := scan()
+	if err != nil {
+		return "", err
+	}
+
+	for _, host := range hosts {
+		if host.mac == id {
+			return host.ip, nil
+		}
+	}
+
+	return "", fmt.Errorf("Device offline")
+}
+
+func PostForm(url, filePath string) error {
+	req := gorequest.New()
+	req.Post(url)
+	req.Type("multipart")
+	req.SendFile(filePath, "firmware.bin", "image")
+
+	log.WithFields(log.Fields{
+		"URL":    req.Url,
+		"Method": req.Method,
+	}).Info("Posting form")
+
+	resp, _, errs := req.End()
+	return handleResp(resp, errs, http.StatusOK)
+}
+
+func activateConnection(connection string) error {
+	cmd := "python activateConnection.py " + connection
+	if err := exec.Command("bash", "-c", cmd).Run(); err != nil {
+		log.WithFields(log.Fields{
+			"Cmd":   cmd,
+			"Error": err,
+		}).Error("Unable to activate connection")
+		return err
+	}
+
+	return nil
+}
+
+func scan() ([]Host, error) {
+	cmd := exec.Command("bash", "-c", "nmap -sP 10.42.0.* -oX scan.txt")
+	cmd.Start()
+
+	done := make(chan error)
+	go func() {
+		done <- cmd.Wait()
+	}()
```

Comment:
The channel will only get cleaned up when there are no longer any references to it, but if the goroutine persists there will be a reference to it from that goro. If there are no readers for done (because we just exited due to timeout) then the goro will be stuck blocked on the write as the channel is unbuffered.

At least that's what I think will happen. ;-)

I think if we make the channel buffered then a write to the channel will always succeed so the goro can finish.